lib/plug/cowboy/handler.ex

```diff
@@ -4,6 +4,9 @@ defmodule Plug.Cowboy.Handler do
   @already_sent {:plug_conn, :sent}
 
   def init(req, {plug, opts}) do
+    start = System.monotonic_time()
+    :telemetry.execute([:plug_cowboy, :handler, :start], %{start_time: System.system_time()})
```

Comment:
I suggest to call it `[:plug, :adapter, :start]`, so all adapters are consistent. We can pass the name of the adapter as metadata. WDYT?

Reply:
Hhmmm... what do you think @GregMefford?

```elixir
:telemetry.execute([:plug, :adapter, :start], %{start_time: System.system_time()}, %{adapter: :plug_cowboy})
```

Reply:
IMO, the name should be whatever the thing actually is, which will usually be similar to the module name where it is fired from (otherwise, why is this module itself not called ‘Plug.Adapter`.

I agree that there’s value in having all Plug adapters emit consistent metrics, but the place to do that would be wherever “Plug itself” could enforce that consistency on all adapters. @bryannaegele and I were looking for some appropriate places in the Plug library itself to do this sort of thing consistently and it was tricky to do because of the way Plug is more of a convention than an actual framework, with the actual code that gets executed being mostly in the user’s code and not in Plug itself.

My concern with calling this `plug.adapter` instead of `plug_cowboy.handler` is that it’s effectively requiring that anyone who wants to make a different adapter should/must implement this event themselves, and if present, it must behave the way this one does. Otherwise, users would be confused why it’s missing/behaving differently.

I’m not sure whether that concern is legitimate or not, but I think it’s best to establish the convention early so that others would follow it: don’t get too fancy about coming up with a generic name for the event, just tell us what the code is literally doing and the user will decide how to assign meaning to it in their context. If I am looking at a trace, either the original or the alternative proposal above would be ok, but the original would lead me quickly to this file to learn more, whereas the more generic alternative would have me looking around in Plug itself and I wouldn’t find what I was looking for there.

Reply:
If we do decide that we like the `plug.adapter.start` naming better, then we’ll need to somehow add that to the docs to explain things like when it should start/end in terms of the Plug API and what the measurement and metadata must include.

Reply:
There is no way in Plug we could enforce this because there is no shared entry point. So if the goal is consistency, the simplest is to actually say all plug adapters have to emit this event. After all, Plug’s goal is to provide consistency across all adapters. If the goal is Cowboy specific metrics, then add telemetry to Cowboy. :)

Reply:
Would it alleviate the discovery problem if called it plug_adapter.request.start? At least people wouldn’t look inside plug itself.

Reply:
I was only proposing adapter consistency because you had originally said “so all adapters are consistent.” I think that it’s fine for this telemetry to be specific to `plug_cowboy` since that’s the library creating it, and some other adapter could have a similar metric with another name and maybe slightly different details that make sense for that adapter.

The names would be listed in the documentation and should be unique. It’s not really a problem of discovery since currently the discovery method is in the docs anyway. I was talking about, for example, if someone saw that event name as a span in a distributed trace, would they be able to easily figure out where it was coming from? If it has the name of it’s associated Hex package as the first atom, for example, it’s easier to namespace and locate where they came from. I don’t think this should be a rule, I’m just saying as a standard philosophy / guiding principal, that might be something to help library authors decide what to call their events.

Separately, I do think we should add various Telemetry hooks in Cowboy as well. If Cowboy itself fires events just before/after the PlugCowboy adapter, and Phoenix (for example) also fires Telemetry around its Endpoint, then perhaps this one being “in between” wouldn’t really add much value besides visibility (via OpenTelemetry-style tracing) into how these things are glued together.

Reply:
So here is an example of where using "plug_cowboy.handler.start" may be impractical.

Phoenix uses its own Cowboy handler. Therefore, should Phoenix call it "phoenix.cowboy_handler.start"? Sure, we could do that, but now when i am swapping between Plug and Phoenix, I need to remember they have different names. Or if I am going to try another adapter, I need to remember to go and change my metrics too.

So I think there is a lot of benefit in having a standardized name. Sure, there is people that may want to know exactly where the event is emitted (and grep would easily find it), but there is also people who will want their metrics to just continue working as they try different things around.

So my take is indeed something like `[:plug_adapter, :request, :start]`.

Reply:
Determined to go with a generic name per @josevalim's suggestion and add the adapter name to the metadata.